java/core/src/main/java/co/worklytics/psoxy/gateway/impl/oauth/ClientCredentialsGrantTokenRequestPayloadBuilder.java

```diff
@@ -69,48 +80,81 @@
+     * @param clientId both subject + issuer of token
+     * @param audience for the token (the endpoint/service being called)
+     * @return JWT assertion as a string
+     */
+    @SneakyThrows
+    String buildJwtAssertion(String clientId, String audience) {
 
+        //https://docs.microsoft.com/en-us/azure/active-directory/develop/active-directory-certificate-credentials
         JsonWebSignature.Header header = new JsonWebSignature.Header();
+        header.setType("JWT");
         header.setAlgorithm("RS256");
-        header.setX509Thumbprint(config.getConfigPropertyOrError(ConfigProperty.PRIVATE_KEY_ID));
-        //header.setType("JWT"); //this IS a JWT, but MSFT examples don't include this
+        header.setX509Thumbprint(encodeKeyId(config.getConfigPropertyOrError(ConfigProperty.PRIVATE_KEY_ID)));
+
 
         JsonWebToken.Payload payload = new JsonWebToken.Payload();
         Instant currentTime = clock.instant();
 
-        payload.setAudience(tokenEndpoint);
+        payload.setAudience(audience);
 
-        payload.setIssuer(oauthClientId);
-        payload.setSubject(oauthClientId);
+        payload.setIssuer(clientId);
+        payload.setSubject(clientId);
 
         payload.setJwtId(uuidGenerator.get().toString()); // MSFT docs say a random GUID, which is what they call UUID
 
         payload.setIssuedAtTimeSeconds(currentTime.getEpochSecond());
         payload.setExpirationTimeSeconds(currentTime.plus(DEFAULT_EXPIRATION_DURATION).getEpochSecond());
 
-        Map<String, String> data = new HashMap<>();
-
-        data.put("grant_type", getGrantType());
-
         String assertion = JsonWebSignature.signUsingRsaSha256(
             getServiceAccountPrivateKey(), jsonFactory, header, payload);
+        return assertion;
+    }
 
-        data.put(PARAM_CLIENT_ASSERTION, assertion);
-        data.put("client_id", oauthClientId);
-        data.put(PARAM_CLIENT_ASSERTION_TYPE, CLIENT_ASSERTION_TYPE_JWT);
-
-        //msft-specific thing
-        config.getConfigPropertyAsOptional(ConfigProperty.RESOURCE)
-            .ifPresent(r -> data.put(PARAM_RESOURCE, r));
-
-        return new UrlEncodedContent(data);
+    //implements encoding X.509 certificate hash (also known as the cert's SHA-1 thumbprint) as a
+    // Base64url string value, which is how MSFT wants it in their JWT assertions
+    @VisibleForTesting
+    String encodeKeyId(String hexKey) {
+        return Base64.getUrlEncoder().encodeToString(hexStringToByteArray(hexKey));
+    }
 
+    //Utility to parse HEX string --> byte[]
+    //TODO: replace with java.util.HexFormat.of().parseHex(s) ... but that's only from Java 17
+    private byte[] hexStringToByteArray(String s) {
+        int len = s.length();
+        byte[] data = new byte[len / 2];
+        for (int i = 0; i < len; i += 2) {
+            data[i / 2] = (byte) ((Character.digit(s.charAt(i), 16) << 4)
+                + Character.digit(s.charAt(i+1), 16));
+        }
+        return data;
```

Comment:
I think we have apache commons-codec transitively (not sure why). It has `Hex.decodeHex` that does this.
Also it need to throw / handle when string length is not even.